src/Remapping/distributed_remapping.jl

```diff
@@ -0,0 +1,297 @@
+    # the vertical column of the interpolated data.
+    local_target_hcoords_bitmask::T3
+
+    # Coefficients (WI1[, WI2]) used for the interpolation. Array of tuples.
+    interpolation_coeffs::T4
+
+    # Local indices the element of each local_target_hcoords in the given topology
+    local_indices::T5
+
+    # Space over which the remapper is defined
+    space::T6
+end
+
+"""
+   Remapper(target_hcoords, target_zcoords, space)
+
+Return a `Remapper` responsible for interpolating any `Field` defined on the given `space`
+to the Cartesian product of `target_hcoords` with `target_zcoords`.
+
+`target_zcoords` can be `nothing` for interpolation on horizontal spaces.
+
+The `Remapper` is designed to not be tied to any particular `Field`. You can use the same
+`Remapper` for any `Field` as long as they are all defined on the same `topology`.
+
+`Remapper` is the main argument to the `interpolate` function.
+
+"""
+function Remapper(target_hcoords, target_zcoords, space)
+
+    comms_ctx = ClimaComms.context(space)
+    pid = ClimaComms.mypid(comms_ctx)
+    FT = Spaces.undertype(space)
+    if hasproperty(space, :horizontal_space)
+        horizontal_topology = space.horizontal_space.topology
+    else
+        horizontal_topology = space.topology
+    end
+
+    is_1d = typeof(horizontal_topology) <: Topologies.IntervalTopology
+
+    # For IntervalTopology, all the points belong to the same process and there's no notion
+    # of containing pid
+    if is_1d
+        # a .== a is an easy way to make a bitmask of the same shape as `a` filled with true
+        local_target_hcoords_bitmask = target_hcoords .== target_hcoords
+    else
+        local_target_hcoords_bitmask =
+            target_hcoords_pid_bitmask(target_hcoords, horizontal_topology, pid)
+    end
+
+    # Extract the coordinate we own (as a MPI process). This will flatten the matrix.
+    local_target_hcoords = target_hcoords[local_target_hcoords_bitmask]
+
+    horz_mesh = horizontal_topology.mesh
+
+    interpolation_coeffs = map(local_target_hcoords) do hcoord
+        helem = Meshes.containing_element(horz_mesh, hcoord)
+        quad = Spaces.quadrature_style(space)
+        quad_points, _ = Spaces.Quadratures.quadrature_points(FT, quad)
+        if is_1d
+            ξ1, = Meshes.reference_coordinates(horz_mesh, helem, hcoord)
+            WI1 = Spaces.Quadratures.interpolation_matrix(
+                SVector(ξ1),
+                quad_points,
+            )
+            return (WI1,)
+        else
+            ξ1, ξ2 = Meshes.reference_coordinates(horz_mesh, helem, hcoord)
+            WI1 = Spaces.Quadratures.interpolation_matrix(
+                SVector(ξ1),
+                quad_points,
+            )
+            WI2 = Spaces.Quadratures.interpolation_matrix(
+                SVector(ξ2),
+                quad_points,
+            )
+            return (WI1, WI2)
+        end
```

Comment:
```suggestion
        ξ = Meshes.reference_coordinates(horz_mesh, helem, hcoord)
        WI1 = Spaces.Quadratures.interpolation_matrix(
            SVector(ξ[1]),
            quad_points,
        )
        if is_1d
            return (WI1,)
        else
            WI2 = Spaces.Quadratures.interpolation_matrix(
                SVector(ξ[2]),
                quad_points,
            )
            return (WI1, WI2)
        end
```

Reply:
eh, above seems clearer.

Reply:
I am going to define a function `interpolation_weights` and dispatch over the function